pkg/kubelet/kubelet.go

```diff
@@ -1452,48 +1460,70 @@ func (kl *Kubelet) SyncPods(allPods []*api.Pod, podSyncTypes map[types.UID]metri
 		return nil
 	}
 
-	// Kill any containers we don't need.
-	killed := []string{}
+	// Kill containers associated with unwanted pods in parallel.
+	type result struct {
+		containers []*docker.Container
+		err        error
+	}
+	ch := make(chan result, len(runningPods))
+	defer close(ch)
+	numWorkers := 0
 	for _, pod := range runningPods {
 		if _, found := desiredPods[pod.ID]; found {
-			// syncPod() will handle this one.
+			// Per-pod workers will handle the desired pods.
 			continue
 		}
-
-		// Kill all the containers in the unidentified pod.
-		for _, c := range pod.Containers {
-			// call the networking plugin for teardown
-			if c.Name == dockertools.PodInfraContainerName {
-				err := kl.networkPlugin.TearDownPod(pod.Namespace, pod.Name, dockertools.DockerID(c.ID))
-				if err != nil {
-					glog.Errorf("Network plugin pre-delete method returned an error: %v", err)
-				}
+		numWorkers++
+		go func(pod *kubecontainer.Pod, ch chan result) {
+			defer func() {
+				// Send the IDs of the containers that we failed to killed.
+				containers, err := kl.getRunningContainersByPod(pod)
+				ch <- result{containers: containers, err: err}
+			}()
+			glog.V(1).Infof("Killing unwanted pod %q", pod.Name)
+			// Stop the containers.
+			err := kl.killPod(*pod)
+			if err != nil {
+				glog.Errorf("Failed killing the pod %q: %v", pod.Name, err)
+				return
 			}
-			glog.V(1).Infof("Killing unwanted container %+v", c)
-			err = kl.killContainer(c)
+			// Remove the pod directory.
+			err = os.RemoveAll(kl.getPodDir(pod.ID))
 			if err != nil {
-				glog.Errorf("Error killing container %+v: %v", c, err)
-			} else {
-				killed = append(killed, string(c.ID))
+				glog.Errorf("Failed removing pod directory for %q", pod.Name)
+				return
 			}
-		}
+		}(pod, ch)
 	}
 
-	running, err := kl.containerManager.GetRunningContainers(killed)
-	if err != nil {
-		glog.Errorf("Failed to poll container state: %v", err)
+	// Aggregate results from the pod killing workers.
+	var errs []string
+	var running []*docker.Container
+	for i := 0; i < numWorkers; i++ {
+		m := <-ch
+		if m.err != nil {
+			errs = append(errs, m.err.Error())
+			continue
+		}
+		running = append(running, m.containers...)
+	}
+	if len(errs) != 0 {
+		err := fmt.Errorf("%s", strings.Join(errs, ", "))
```

Comment:
Consider using errors.Aggregate so we have similar error messages:

https://github.com/yujuhong/kubernetes/blob/para_cleanup/pkg/util/errors/errors.go#L23